src/build_vehicle_gui.cpp

```diff
@@ -269,20 +269,25 @@ static bool EnginePowerVsRunningCostSorter(const EngineID &a, const EngineID &b)
 {
 	const Engine *e_a = Engine::Get(a);
 	const Engine *e_b = Engine::Get(b);
+	uint p_a = e_a->GetPower();
+	uint p_b = e_b->GetPower();
+	Money r_a = e_a->GetRunningCost();
+	Money r_b = e_b->GetRunningCost();
+	/* Check if running cost is zero in one or both engines */
+	if (r_a == 0) {
+		if (r_b == 0) {
+			if (p_a == p_b) return EngineNumberSorter(a,b);
+			return _engine_sort_direction ? p_a >= p_b : p_a < p_b;
+		}
+		return !_engine_sort_direction;
+	}
+	if (r_b == 0 ) return _engine_sort_direction;
 
-	/* Here we are using a few tricks to get the right sort.
-	 * We want power/running cost, but since we usually got higher running cost than power and we store the result in an int,
-	 * we will actually calculate cunning cost/power (to make it more than 1).
-	 * Because of this, the return value have to be reversed as well and we return b - a instead of a - b.
-	 * Another thing is that both power and running costs should be doubled for multiheaded engines.
-	 * Since it would be multiplying with 2 in both numerator and denominator, it will even themselves out and we skip checking for multiheaded. */
-	Money va = (e_a->GetRunningCost()) / max(1U, (uint)e_a->GetPower());
-	Money vb = (e_b->GetRunningCost()) / max(1U, (uint)e_b->GetPower());
-	int r = ClampToI32(vb - va);
-
+	float va = p_a / r_a;
+	float vb = p_b / r_b;
```

Comment:
Isn't this an integer division anyway? Should've broken all the normal sorting.

Reply:
You are right. Forgot to cast (float) on both integers.

Reply:
You'll lose some precision in that case and that == later doesn't make much sense for floats. Though it's probably fine since it's a gui thing anyway it still needs to be mentioned in a comment imo.

Reply:
It was for the case that both where zero.

Reply:
Also could use doubles for more precision. But i am not sure if that's too much.